File: LeetCode75/array_and_string/greatest_common_divisor_of_strings/greatest_common_divisor_of_strings.cpp
#include <string>
#include <string_view>

using std::string;

class Solution
{
  public:
    string gcdOfStrings(string str1, string str2) const
    {
        size_t longest{0};
        size_t shorter_len{std::min(str1.length(), str2.length())};
        while (longest < shorter_len && str1[longest] == str2[longest])
            longest++;

        while (longest != 0)
        {
            if (str1.length() % longest == 0 && str2.length() % longest == 0)
            {
                std::string_view sv{std::string_view(str1).substr(0, longest)};
                if (divides(sv, str1) && divides(sv, str2))
                    return string(sv);
            }
            longest--;
        }
        return "";
    }

  private:
    bool divides(std::string_view shorter,
                 std::string_view longer) const noexcept
    {
        auto s_iter{shorter.begin()};
        auto l_iter{longer.begin()};
        while (l_iter != longer.end())
        {
            if (s_iter >= shorter.end())
                s_iter = shorter.begin();
            if (*l_iter != *s_iter)
                return false;
            ++s_iter;
            ++l_iter;
        }
        return s_iter == shorter.end() && l_iter == longer.end();
    }
};

#include <iostream>

int main()
{
    Solution solution;

    string a{"ABABAB"};
    string b{"ABAB"};

    std::cout << solution.gcdOfStrings(a, b) << "\n";
}

File: LeetCode75/array_and_string/greatest_common_divisor_of_strings/greatest_common_divisor_of_strings.py
from typing import List


class Solution:
    def gcdOfStrings(self, str1: str, str2: str) -> str:
        def divides(short: str, long: str) -> bool:
            return short * (len(long) // len(short)) == long

        divisor_len = 0
        len_max = min(len(str1), len(str2))
        for i in range(len_max):
            if str1[i] == str2[i]:
                divisor_len += 1
            else:
                break
        while divisor_len > 0:
            if len(str1) % divisor_len == 0 and len(str2) % divisor_len == 0:
                canditate = str1[:divisor_len]
                if divides(canditate, str1) and divides(canditate, str2):
                    return str1[0:divisor_len]
            divisor_len -= 1
        return ""


if __name__ == "__main__":
    solution = Solution()
    print(solution.gcdOfStrings("ABC", "ABCDEF"))
